Sources/NIO/SelectorUring.swift

```diff
@@ -0,0 +1,359 @@
+private struct UringFilterSet: OptionSet, Equatable {
+    typealias RawValue = UInt8
+
+    let rawValue: RawValue
+
+    static let _none = UringFilterSet([])
+    static let hangup = UringFilterSet(rawValue: 1 << 0)
+    static let readHangup = UringFilterSet(rawValue: 1 << 1)
+    static let input = UringFilterSet(rawValue: 1 << 2)
+    static let output = UringFilterSet(rawValue: 1 << 3)
+    static let error = UringFilterSet(rawValue: 1 << 4)
+
+    init(rawValue: RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+extension UringFilterSet {
+    /// Convert NIO's `SelectorEventSet` set to a `UringFilterSet`
+    init(selectorEventSet: SelectorEventSet) {
+        var thing: UringFilterSet = [.error, .hangup]
+        if selectorEventSet.contains(.read) {
+            thing.formUnion(.input)
+        }
+        if selectorEventSet.contains(.write) {
+            thing.formUnion(.output)
+        }
+        if selectorEventSet.contains(.readEOF) {
+            thing.formUnion(.readHangup)
+        }
+        self = thing
+    }
+}
+
+extension SelectorEventSet {
+    var uringEventSet: UInt32 {
+        assert(self != ._none)
+        // POLLERR | POLLHUP is always set unconditionally anyway but it's easier to understand if we explicitly ask.
+        var filter: UInt32 = Uring.POLLERR | Uring.POLLHUP
+        let uringFilters = UringFilterSet(selectorEventSet: self)
+        if uringFilters.contains(.input) {
+            filter |= Uring.POLLIN
+        }
+        if uringFilters.contains(.output) {
+            filter |= Uring.POLLOUT
+        }
+        if uringFilters.contains(.readHangup) {
+            filter |= Uring.POLLRDHUP
+        }
+        assert(filter & Uring.POLLHUP != 0) // both of these are reported
+        assert(filter & Uring.POLLERR != 0) // always and can't be masked.
+        return filter
+    }
+
+    fileprivate init(uringEvent: UInt32) {
+        var selectorEventSet: SelectorEventSet = ._none
+        if uringEvent & Uring.POLLIN != 0 {
+            selectorEventSet.formUnion(.read)
+        }
+        if uringEvent & Uring.POLLOUT != 0 {
+            selectorEventSet.formUnion(.write)
+        }
+        if uringEvent & Uring.POLLRDHUP != 0 {
+            selectorEventSet.formUnion(.readEOF)
+        }
+        if uringEvent & Uring.POLLHUP != 0 || uringEvent & Uring.POLLERR != 0 {
+            selectorEventSet.formUnion(.reset)
+        }
+        self = selectorEventSet
+    }
+}
+
+extension Selector: _SelectorBackendProtocol {
+    internal func _debugPrint(_ s : @autoclosure () -> String) {
+        #if SWIFTNIO_IO_URING_DEBUG_SELECTOR
+        print("S [\(NIOThread.current)] " + s())
+        #endif
+    }
+
+    func initialiseState0() throws {
+        try ring.io_uring_queue_init()
+        self.selectorFD = ring.fd()
+
+        // eventfd are always singleshot and re-register each time around
+        // as certain use cases of nio seems to generate tons of wakeups
```

Comment:
Yup, this "tons of wakeups" is not what we want to do at the moment, but it's where we're at. We should investigate getting rid of them. In the fullness of time we'll be at risk of some redundant wakeups because we'll want to move to an MPSC queue, but right now when we have a big dumb lock we can absolutely avoid spurious wakes.

Reply:
Yeah, wasn't perhaps the most accurate term, updated the comment. MPSC seems like an interesting direction.